x/denommetadata/keeper/msg_server_test.go

```diff
@@ -1,49 +1,178 @@
+				SenderAddress: senderAddress,
+				TokenMetadata: denomMetadata,
+			},
+			hooks:            &mockERC20Hook{},
+			expectHookCalled: true,
+		}, {
+			name: "permission error",
+			msg: &types.MsgCreateDenomMetadata{
+				SenderAddress: senderAddress,
+				TokenMetadata: denomMetadata,
+			},
+			malleate: func() {
+				initialParams := types.DefaultParams()
+				initialParams.AllowedAddresses = []string{}
+				suite.k.SetParams(suite.ctx, initialParams)
+			},
+			hooks:            &mockERC20Hook{},
+			expectHookCalled: false,
+			expectErr:        types.ErrNoPermission.Error(),
+		}, {
+			name: "denom already exists",
+			msg: &types.MsgCreateDenomMetadata{
+				SenderAddress: senderAddress,
+				TokenMetadata: denomMetadata,
+			},
+			malleate: func() {
+				msg := &types.MsgCreateDenomMetadata{
+					SenderAddress: senderAddress,
+					TokenMetadata: denomMetadata,
+				}
+				_, err := suite.msgServer.CreateDenomMetadata(suite.ctx, msg)
+				suite.Require().NoError(err, "CreateDenomMetadata() error")
+			},
+			hooks:            &mockERC20Hook{},
+			expectHookCalled: false,
+			expectErr:        types.ErrDenomAlreadyExists.Error(),
+		}, {
+			name: "invalid denom units",
+			msg: &types.MsgCreateDenomMetadata{
+				SenderAddress: senderAddress,
+				TokenMetadata: banktypes.Metadata{
+					Description: "ATOM IBC",
+					Base:        ibcBase,
+					DenomUnits: []*banktypes.DenomUnit{
+						{Denom: ibcBase, Exponent: 18},
+						{Denom: "ATOM", Exponent: 0},
+					},
+					Name:    "ATOM channel-0",
+					Symbol:  "ibcATOM-0",
+					Display: ibcBase,
+				},
+			},
+			hooks:            &mockERC20Hook{},
+			expectHookCalled: false,
+			expectErr:        fmt.Sprintf("the exponent for base denomination unit %s must be 0", ibcBase),
+		}, {
+			name: "failed to create erc20 contract",
+			msg: &types.MsgCreateDenomMetadata{
+				SenderAddress: senderAddress,
+				TokenMetadata: denomMetadata,
+			},
+			hooks: &mockERC20Hook{
+				err: fmt.Errorf("failed to deploy the erc20 contract for the IBC coin"),
+			},
+			expectHookCalled: false,
+			expectErr:        "error in after denom metadata creation hook: failed to deploy the erc20 contract for the IBC coin",
+		},
+	}
+
+	for _, tc := range cases {
+		suite.Run(tc.name, func() {
+			// reset the test state
+			suite.setupTest(tc.hooks)
+
+			if tc.malleate != nil {
+				tc.malleate()
+			}
+
+			_, err := suite.msgServer.CreateDenomMetadata(suite.ctx, tc.msg)
+			if tc.expectErr != "" {
+				suite.Require().ErrorContains(err, tc.expectErr)
+				return
+			}
+			suite.Require().NoError(err, "CreateDenomMetadata() error")
+
+			// check if the denom metadata was added
+			_, found := suite.app.BankKeeper.GetDenomMetaData(suite.ctx, ibcBase)
+			suite.Require().True(found, "denom metadata should exist")
+			suite.Require().Equal(tc.expectHookCalled, tc.hooks.createCalled, "after denom metadata creation hook should be called")
+		})
+	}
+}
+
+type mockERC20Hook struct {
```

Comment:
next time u can use `mockery` which generates the mocks and provides assertion interface
instead of custom mock struct